### ISSUE TITLE
assert(!array_is_list()) changes type to be maybe list

### ISSUE THREAD

**Issue by othercorey** [vimeo/psalm#9348]
This is a strange issue that seems straight-forward to reproduce in psalm.dev, but does not trigger there.

```
ERROR: InvalidArgument - src/ORM/Table.php:2590:30 - Argument 1 of Cake\ORM\Query\SelectQuery::applyOptions expects array<string, mixed>, but array<int<0, max>|string, mixed> provided (see https://psalm.dev/004)
        $query->applyOptions($options);
```

A parameter of type `array<string, mixed` is passed to another parameter of the same type `array<string, mixed>`.

However, when introducing an assert that it is associative, it seems to think the array could now be a list.

Same code in psalm.dev does not throw an error: https://psalm.dev/r/c7b80db58a

The function starts with this code so not sure what else could be affecting it:
```php
    public function callFinder(string $type, SelectQuery $query, array $options = []): SelectQuery
    {
        assert(empty($options) || !array_is_list($options), 'Finder options should be an associative array not a list');

        $query->applyOptions($options);
```

**Comment by weirdan** [comment]
@othercorey you omitted `@` in `@param`, that's why you couldn't reproduce it on psalm.dev. Here's the fixed reproducer: https://psalm.dev/r/715994eafd

**Comment by othercorey** [comment]
Thanks. I updated the description with your example so it's not confusing to read.
